File: backend/src/services/memoryAllowance.ts
import type { MemoryPlan, PrismaClient } from "@prisma/client";
import { HttpError } from "../lib/httpError.js";
import { MEMORY_UPGRADE_WEEKLY_CENTS } from "./subscriptionCatalog.js";

/**
 * **Les limites de souvenirs** : ce qu'un compte peut raconter dans le mois, et
 * ce que ça coûte de raconter plus.
 *
 * À ne pas confondre avec les **étapes offertes** (`quota.ts`), qui sont le
 * palier d'entrée : trois étapes, une fois, puis l'abonnement. Les limites de
 * souvenirs sont le budget **hebdomadaire** de quelqu'un qui raconte **déjà** —
 * un abonné. Elles se rechargent toutes les semaines, et se relèvent contre
 * 3,99 €/semaine.
 *
 * **La semaine, parce que tout le produit est à la semaine** (Hugo,
 * 17/09/2026) : l'abonnement se facture ainsi, un voyage se compte ainsi, et
 * l'extension est une option du même produit. Un plafond mensuel derrière un
 * prix hebdomadaire aurait annoncé quatre fois le montant affiché.
 *
 * ⚠️ **On ne dit jamais « jetons » ni « tokens » dans l'app** (Hugo,
 * 16/09/2026). Le mot est « souvenirs », et l'unité aussi : l'app annonce
 * « 1 240 souvenirs sur 3 000 », pas un pourcentage de quota. Ce fichier est le
 * seul endroit où la conversion se fait ; l'app ne compte rien.
 *
 * **Pourquoi un vocal coûte plus qu'un message.** Parce qu'il coûte plus : une
 * minute de vocal passe par la transcription, la rédaction et la relecture, là
 * où un message écrit n'appelle que les deux dernières sur dix fois moins de
 * texte. Le rapport de dix est un ordre de grandeur, pas une mesure — il est à
 * réétalonner sur les factures OpenAI et Anthropic d'un mois plein, et c'est
 * pour ça qu'il vit ici, en une constante, plutôt que dispersé dans les routes.
 */

/** Ce qu'un message écrit consomme. L'unité de compte. */
export const TEXT_MEMORY_COST = 1;

/**
 * Ce qu'une **minute entamée** de vocal consomme.
 *
 * Entamée et non écoulée : une minute et une seconde en coûtent deux. C'est la
 * règle la plus facile à expliquer, et la seule qui ne récompense pas le
 * découpage d'un vocal en morceaux de 59 secondes.
 */
export const VOICE_MEMORY_COST_PER_MINUTE = 10;

/** Ce que chaque palier ouvre par semaine. */
export const MEMORY_ALLOWANCE: Record<MemoryPlan, number> = {
  // **200 minutes de vocal par semaine**, ou 2 000 messages — près de
  // 30 minutes de vocal par jour. Volontairement haut : cette limite n'est pas
  // un levier commercial, c'est un garde-fou contre l'usage qui coûterait plus
  // cher que l'abonnement. Un voyageur bavard qui raconte 20 minutes par jour
  // en consomme 1 400 : il ne doit jamais la voir bouger.
  included: 2_000,
  // Quatre fois plus, pour un voyage raconté à plusieurs, tous les jours.
  extended: 8_000,
};

/**
 * La durée d'une période, en jours. **Une semaine glissante**, comme la
 * facturation — voir `MEMORY_UPGRADE_WEEKLY_CENTS`.
 */
const PERIOD_DAYS = 7;

/** Le prix hebdomadaire du palier étendu, en centimes. Réexporté pour l'API. */
export const MEMORY_UPGRADE_CENTS = MEMORY_UPGRADE_WEEKLY_CENTS;

export type MemorySnapshot = {
  plan: MemoryPlan;
  used: number;
  allowance: number;
  /** Le jour où le budget se remet à zéro. */
  renewsOn: Date;
  upgradeWeeklyPrice: number;
};

/**
 * Ce qu'une durée de vocal coûte, en souvenirs.
 *
 * Une durée absente coûte une minute : un vocal dont on ignore la durée existe
 * quand même, et ne rien décompter reviendrait à offrir l'usage le plus cher à
 * qui ne transmet pas la métadonnée.
 */
export function voiceCost(durationSeconds: number | null | undefined): number {
  const seconds = durationSeconds ?? 60;
  const minutes = Math.max(1, Math.ceil(seconds / 60));
  return minutes * VOICE_MEMORY_COST_PER_MINUTE;
}

/**
 * Le début de période à retenir : celui qu'on a, ou aujourd'hui si la semaine
 * est écoulée.
 *
 * **Glissante et non calendaire** : un compte ouvert un samedi aurait sinon un
 * jour de budget pour sa première semaine, ce qui se lit comme un bogue et ne
 * se rattrape par aucune explication.
 */
function currentPeriodStart(start: Date, now: Date): Date {
  const elapsed = now.getTime() - start.getTime();
  const period = PERIOD_DAYS * 24 * 60 * 60 * 1000;
  if (elapsed < period) return start;

  // On saute les périodes entières écoulées d'un coup plutôt que de repartir de
  // maintenant : le jour de renouvellement reste le même d'une semaine à
  // l'autre, même après un mois sans ouvrir l'app.
  const skipped = Math.floor(elapsed / period);
  return new Date(start.getTime() + skipped * period);
}

/** Le jour où le budget se remet à zéro. */
function periodEnd(start: Date): Date {
  return new Date(start.getTime() + PERIOD_DAYS * 24 * 60 * 60 * 1000);
}

/**
 * Où en est un compte, **période remise à jour si elle a expiré**.
 *
 * La remise à zéro a lieu à la lecture et non par une tâche planifiée : un
 * budget hebdomadaire n'a pas besoin d'être exact à la seconde, il a besoin
 * d'être juste au moment où quelqu'un le regarde ou le consomme. Une tâche de
 * plus pour ça serait une pièce mobile de plus à surveiller.
 */
export async function readMemoryAllowance(
  prisma: PrismaClient,
  accountId: string,
): Promise<MemorySnapshot> {
  const account = await prisma.account.findUniqueOrThrow({
    where: { id: accountId },
    select: { memoryPlan: true, memoryUsed: true, memoryPeriodStart: true },
  });

  const now = new Date();
  const start = currentPeriodStart(account.memoryPeriodStart, now);
  const isNewPeriod = start.getTime() !== account.memoryPeriodStart.getTime();

  if (isNewPeriod) {
    await prisma.account.update({
      where: { id: accountId },
      data: { memoryPeriodStart: start, memoryUsed: 0 },
    });
  }

  return {
    plan: account.memoryPlan,
    used: isNewPeriod ? 0 : account.memoryUsed,
    allowance: MEMORY_ALLOWANCE[account.memoryPlan],
    renewsOn: periodEnd(start),
    upgradeWeeklyPrice: MEMORY_UPGRADE_CENTS,
  };
}

/**
 * Décompte, et refuse quand il ne reste rien.
 *
 * **Le refus porte un code que l'app sait lire** (`memory_limit_reached`) :
 * c'est lui qui ouvre la feuille d'extension plutôt qu'un bandeau d'erreur. Un
 * 403 muet aurait envoyé quelqu'un au support pour une limite qui s'achète en
 * deux gestes.
 *
 * On décompte **après** avoir vérifié, dans la même écriture : deux vocaux
 * envoyés en même temps ne doivent pas passer tous les deux sur le dernier
 * souvenir disponible. `increment` fait l'addition côté base, donc sans course.
 */
export async function consumeMemory(
  prisma: PrismaClient,
  accountId: string,
  cost: number,
): Promise<void> {
  if (cost <= 0) return;

  const snapshot = await readMemoryAllowance(prisma, accountId);
  if (snapshot.used + cost > snapshot.allowance) {
    throw new HttpError(
      403,
      "Tu as atteint tes limites de souvenirs pour cette semaine. Étends-les pour continuer à raconter.",
      "memory_limit_reached",
    );
  }

  await prisma.account.update({
    where: { id: accountId },
    data: { memoryUsed: { increment: cost } },
  });
}

/**
 * Passe au palier étendu.
 *
 * ⚠️ **Aucun encaissement ici.** Comme l'abonnement lui-même, l'extension est
 * un service numérique : Apple impose l'achat intégré, et c'est StoreKit qui
 * portera la transaction. Cette route pose le palier, ce qui permet de
 * dérouler le parcours de bout en bout dans l'app ; le jour où StoreKit est
 * branché, c'est son reçu qui l'appelle.
 */
export async function setMemoryPlan(
  prisma: PrismaClient,
  accountId: string,
  plan: MemoryPlan,
): Promise<MemorySnapshot> {
  await prisma.account.update({ where: { id: accountId }, data: { memoryPlan: plan } });
  return readMemoryAllowance(prisma, accountId);
}

File: ios/Modules/Sources/MemoBookCore/MemoryAllowance.swift
import Foundation

// Les **limites de souvenirs** : ce qu'un compte peut raconter dans la semaine.
//
// ⚠️ **Le mot « jeton » n'apparaît nulle part**, et c'est une consigne, pas une
// préférence (Hugo, 16/09/2026). L'unité s'appelle un **souvenir**, et l'app
// compte en souvenirs : « 1 240 sur 3 000 ». Personne n'ouvre MemoBook pour
// gérer un budget de calcul.
//
// Le barème lui-même vit côté serveur (`services/memoryAllowance.ts`) et
// **voyage avec le solde** : ce que coûte un message et ce que coûte une minute
// de vocal sont dans la réponse, pas dans le binaire. La feuille d'extension a
// besoin de les écrire pour expliquer pourquoi un vocal pèse plus — et une
// seconde table de prix dans l'app serait une seconde vérité à tenir d'accord.

/// Le palier de limites d'un compte.
public enum MemoryPlan: String, Codable, Sendable, Hashable, CaseIterable, Identifiable {
    /// Ce qui est compris. Le palier de tout le monde.
    case included
    /// Le palier étendu, à 3,99 €/semaine.
    case extended

    public var id: String { rawValue }

    /// Le nom du palier, tel que la feuille l'écrit.
    public var title: String {
        switch self {
        case .included: "Limites comprises"
        case .extended: "Limites étendues"
        }
    }
}

/// Où en est un compte de ses limites de souvenirs.
///
/// **Servi avec les réglages du voyage** — comme la cagnotte, et pour la même
/// raison : ça appartient au compte, mais c'est le seul écran qui le montre.
public struct MemoryAllowance: Codable, Sendable, Hashable {
    public var plan: MemoryPlan
    /// Ce qui a été consommé dans la période en cours.
    public var used: Int
    /// Ce que la période ouvre.
    public var allowance: Int
    /// Le jour où le compteur repart à zéro.
    public var renewsOn: Date?
    /// Ce que coûte le palier étendu, **par semaine**.
    ///
    /// Hebdomadaire comme l'abonnement, et pour la même raison : c'est une
    /// option du même produit, pas une seconde offre (Hugo, 17/09/2026).
    public var upgradeWeeklyPrice: Decimal
    /// Ce qu'un message écrit consomme.
    public var textCost: Int
    /// Ce qu'une **minute entamée** de vocal consomme.
    public var voiceCostPerMinute: Int

    public init(
        plan: MemoryPlan = .included,
        used: Int = 0,
        allowance: Int = 2_000,
        renewsOn: Date? = nil,
        upgradeWeeklyPrice: Decimal = 3.99,
        textCost: Int = 1,
        voiceCostPerMinute: Int = 10
    ) {
        self.plan = plan
        self.used = used
        self.allowance = allowance
        self.renewsOn = renewsOn
        self.upgradeWeeklyPrice = upgradeWeeklyPrice
        self.textCost = textCost
        self.voiceCostPerMinute = voiceCostPerMinute
    }

    /// Décodage tolérant : un serveur qui ne sert pas encore le barème ne doit
    /// pas faire échouer tout l'écran des réglages. On retombe alors sur les
    /// valeurs par défaut, qui sont celles du catalogue.
    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        plan = try container.decodeIfPresent(MemoryPlan.self, forKey: .plan) ?? .included
        used = try container.decodeIfPresent(Int.self, forKey: .used) ?? 0
        allowance = try container.decodeIfPresent(Int.self, forKey: .allowance) ?? 2_000
        renewsOn = try container.decodeIfPresent(Date.self, forKey: .renewsOn)
        upgradeWeeklyPrice =
            try container.decodeIfPresent(Decimal.self, forKey: .upgradeWeeklyPrice) ?? 3.99
        textCost = try container.decodeIfPresent(Int.self, forKey: .textCost) ?? 1
        voiceCostPerMinute =
            try container.decodeIfPresent(Int.self, forKey: .voiceCostPerMinute) ?? 10
    }

    // MARK: Ce que la jauge lit

    /// Ce qu'il reste. Jamais négatif : un dépassement — un barème réétalonné
    /// entre deux périodes — se lit « 0 restant », pas « -40 ».
    public var remaining: Int { max(0, allowance - used) }

    /// Où en est la jauge, entre 0 et 1.
    public var fraction: Double {
        guard allowance > 0 else { return 0 }
        return min(1, Double(used) / Double(allowance))
    }

    /// **Le seuil à partir duquel on le dit.** En dessous, la ligne affiche son
    /// solde sans rien réclamer : cette limite est un garde-fou, pas un levier,
    /// et quelqu'un qui en a consommé un tiers n'a rien à décider.
    public static let warningFraction = 0.8

    /// Il est temps de proposer d'étendre.
    public var isRunningLow: Bool {
        plan == .included && fraction >= Self.warningFraction
    }

    /// Plus rien. Le micro et l'envoi se ferment — le serveur refuse de son
    /// côté avec `memory_limit_reached`.
    public var isExhausted: Bool { remaining == 0 }

    /// Ce que le palier étendu ouvrirait. Un multiple et non un nombre écrit à
    /// la main : si le barème bouge, la phrase suit.
    public func multiplier(over extendedAllowance: Int) -> Int {
        guard allowance > 0 else { return 1 }
        return max(1, extendedAllowance / allowance)
    }
}
